plugins/aggregators/histogram/histogram.go

```diff
@@ -167,18 +180,35 @@ func (h *HistogramAggregator) groupFieldsByBuckets(
 	tags map[string]string,
 	counts []int64,
 ) {
-	count := int64(0)
-	for index, bucket := range h.getBuckets(name, field) {
-		count += counts[index]
+	if h.Cumulative {
+		count := int64(0)
+		for index, bucket := range h.getBuckets(name, field) {
+			count += counts[index]
 
-		tags[bucketTag] = strconv.FormatFloat(bucket, 'f', -1, 64)
-		h.groupField(metricsWithGroupedFields, name, field, count, copyTags(tags))
-	}
+			tags[bucketRightTag] = strconv.FormatFloat(bucket, 'f', -1, 64)
+			h.groupField(metricsWithGroupedFields, name, field, count, copyTags(tags))
+		}
 
-	count += counts[len(counts)-1]
-	tags[bucketTag] = bucketInf
+		count += counts[len(counts)-1]
+		tags[bucketRightTag] = bucketPosInf
 
-	h.groupField(metricsWithGroupedFields, name, field, count, tags)
+		h.groupField(metricsWithGroupedFields, name, field, count, tags)
+	} else {
+		buckets := h.getBuckets(name, field)
+		for index, count := range counts {
+			tags[bucketLeftTag] = bucketNegInf
+			if index > 0 {
+				tags[bucketLeftTag] = strconv.FormatFloat(buckets[index-1], 'f', -1, 64)
+			}
```

Comment:
I think all we should add is non-cumulative buckets, can't we just stick the raw counts into the bucket instead of incrementing?  The bucket specifiers in the configuration are defined as the right borders, so by definition nothing should go into the -Inf bucket, negative values should be included in the smallest bucket.

Reply:
Hello @danielnelson, what exactly do you mean by "stick the raw counts into the bucket instead of incrementing"?

> The bucket specifiers in the configuration are defined as the right borders, so by definition nothing should go into the -Inf bucket, negative values should be included in the smallest bucket.

When you say they should go "into the XY bucket", you're meaning the bucket with right border XY, right? Because that's just how my implementation works, I only use the specified bucket borders and +Inf as right bucket borders - the `le` tag (which is the right bucket border) has the same values as in the cumulative case. However, in the non-cumulative case, buckets not only have an upper bound but also a lower bound. I wanted to also include that information with the `gt` tag as it can be quite difficult in tools like grafana to get information from the next-lower bucket to receive the left border of the current bucket. I also find this more concrete, as a metric with tags `gt=0.0, le=1.0` simply defines the mathematical interval `]0, 10]`.  (In the cumulative case, intervals are always `]-Inf, 0]`, `]-Inf, 10]`, ...)

Reply:
I see now, was mixed up by the change in what is being iterated over, buckets vs counts, and I didn't notice that -Inf is for a new tag.

What I'm wondering now is if it makes sense to merge these loops, they are mostly doing the same thing only the non-cumulative doesn't add to count and adds the extra tag.  Something like this (untested):

```diff
+ lowerBound := "-Inf"
count := int64(0)
for index, bucket := range h.getBuckets(name, field) {
+	if !h.Cumulative {
+		count = 0
+		tags[bucketLeftTag] = lowerBound
+		lowerBound = strconv.FormatFloat(bucket, 'f', -1, 64)
+	}

	count += counts[index]
	tags[bucketRightTag] = strconv.FormatFloat(bucket, 'f', -1, 64)
	h.groupField(metricsWithGroupedFields, name, field, count, copyTags(tags))
}
```